File: python/realtime/audio_out_ws.py
from __future__ import annotations

import asyncio
import base64
import json
import logging
import time
from dataclasses import dataclass, field

from . import audio_defaults
from .audio_out import OutboundPushError

log = logging.getLogger("realtime.audio_out_ws")

KOKORO_HZ = audio_defaults.TTS_SAMPLE_RATE
FRAME_MS = audio_defaults.FRAME_MS

def _wire_for(format: str) -> tuple[str, int]:
    table = {
        "pcm16_8k": ("pcm16le", 8_000),
        "pcm16_16k": ("pcm16le", 16_000),
        "pcm16_24k": ("pcm16le", 24_000),
        "pcm16_44k1": ("pcm16le", 44_100),
        "pcm16_48k": ("pcm16le", 48_000),
        "g711_ulaw": ("ulaw", 8_000),
        "g711_alaw": ("alaw", 8_000),
    }
    return table.get(format, ("pcm16le", 24_000))

@dataclass
class WsAudioPacer:
    ws_send: asyncio.Queue
    id_event_factory: callable
    played_ms_ref: list[int]
    codec: str
    sample_rate: int
    frame_samples: int
    frames_written: int = 0
    _start: float | None = None
    carry: list[float] = field(default_factory=list)
    last_sample: float = 0.0
    src_position: float = 0.0

    @classmethod
    def start(
        cls,
        ws_send: asyncio.Queue,
        id_event_factory: callable,
        played_ms_ref: list[int],
        format: str,
    ) -> "WsAudioPacer":
        codec, sr = _wire_for(format)
        frame_samples = sr * FRAME_MS // 1000
        return cls(
            ws_send=ws_send,
            id_event_factory=id_event_factory,
            played_ms_ref=played_ms_ref,
            codec=codec,
            sample_rate=sr,
            frame_samples=frame_samples,
        )

    async def play(self, audio_24k_samples: list[float]) -> None:
        if not audio_24k_samples:
            return
        if self.sample_rate == KOKORO_HZ:
            resampled = list(audio_24k_samples)
        else:
            resampled = self._linear_resample(audio_24k_samples)
        self.carry.extend(resampled)
        while len(self.carry) >= self.frame_samples:
            await self._emit_one_frame()

    async def flush(self) -> None:
        if not self.carry:
            return
        while len(self.carry) < self.frame_samples:
            self.carry.append(0.0)
        await self._emit_one_frame()

    async def _emit_one_frame(self) -> None:
        frame = self.carry[: self.frame_samples]
        self.carry = self.carry[self.frame_samples :]
        if self.codec == "pcm16le":
            buf = bytearray(self.frame_samples * 2)
            for i, s in enumerate(frame):
                v = int(max(-1.0, min(1.0, s)) * 32_767.0)
                if v < 0:
                    v += 0x10000
                buf[i * 2] = v & 0xFF
                buf[i * 2 + 1] = (v >> 8) & 0xFF
            payload = bytes(buf)
        elif self.codec in ("ulaw", "alaw"):
            from audio.g711 import alaw_encode_sample, ulaw_encode_sample

            enc = ulaw_encode_sample if self.codec == "ulaw" else alaw_encode_sample
            buf = bytearray(self.frame_samples)
            for i, s in enumerate(frame):
                v = int(max(-1.0, min(1.0, s)) * 32_767.0)
                buf[i] = enc(v)
            payload = bytes(buf)
        else:
            return

        event = {
            "event_id": self.id_event_factory(),
            "type": "response.output_audio.delta",
            "delta": base64.b64encode(payload).decode("ascii"),
        }
        try:
            text = json.dumps(event, separators=(",", ":"))
        except (TypeError, ValueError) as err:
            log.warning("audio.delta json serialize failed: %s", err)
            return
        try:
            await self.ws_send.put(text)
        except Exception as err:
            log.warning("ws writer dropped while sending audio.delta: %s", err)
            return

        if self._start is None:
            self._start = time.monotonic()
        self.frames_written += 1
        if self.played_ms_ref:
            self.played_ms_ref[0] = self.frames_written * FRAME_MS
        target = self._start + (self.frames_written * FRAME_MS) / 1000.0
        now = time.monotonic()
        if target > now:
            await asyncio.sleep(target - now)

    def _linear_resample(self, src: list[float]) -> list[float]:
        if not src:
            return []
        ratio = self.sample_rate / float(KOKORO_HZ)
        out: list[float] = []
        pos = self.src_position
        step = 1.0 / ratio
        last = self.last_sample
        n = len(src)
        while pos < n:
            lo = int(pos)
            hi = min(lo + 1, n - 1)
            frac = pos - lo
            if lo >= n:
                s = last
            else:
                a = src[lo]
                b = src[hi]
                s = a + (b - a) * frac
            out.append(s)
            last = s
            pos += step
        self.src_position = pos - n
        self.last_sample = last
        return out

@dataclass
class AudioPacer:
    webrtc: object | None = None
    websocket: WsAudioPacer | None = None

    @classmethod
    def from_webrtc(cls, p) -> "AudioPacer":
        return cls(webrtc=p)

    @classmethod
    def from_websocket(cls, p: WsAudioPacer) -> "AudioPacer":
        return cls(websocket=p)

    async def play(self, audio_24k_samples) -> None:
        if self.webrtc is not None:
            await self.webrtc.play(audio_24k_samples)
            return
        if self.websocket is not None:
            await self.websocket.play(audio_24k_samples)

    async def flush(self) -> None:
        if self.webrtc is not None:
            await self.webrtc.flush()
            return
        if self.websocket is not None:
            await self.websocket.flush()

__all__ = [
    "AudioPacer",
    "OutboundPushError",
    "WsAudioPacer",
]

File: rust/src/realtime/audio_out_ws.rs
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};

use anyhow::Result;
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine as _;
use serde_json::json;
use tokio::sync::mpsc;
use tracing::warn;

use super::audio_out::OutboundPushError;
use super::wire::EventSeq;
use crate::audio::g711;
use crate::types::MonoF32At24k;

const KOKORO_HZ: usize = 24_000;
const FRAME_MS: usize = 20;
const OUTPUT_INDEX: u64 = 0;
const CONTENT_INDEX: u64 = 0;

#[derive(Debug, Clone, Copy)]
enum WireCodec {
    Pcm16Le,
    Ulaw,
    Alaw,
}

fn wire_for(format: &str) -> (WireCodec, usize) {
    match format {
        "pcm16_8k" => (WireCodec::Pcm16Le, 8_000),
        "pcm16_16k" => (WireCodec::Pcm16Le, 16_000),
        "pcm16_24k" => (WireCodec::Pcm16Le, 24_000),
        "pcm16_44k1" => (WireCodec::Pcm16Le, 44_100),
        "pcm16_48k" => (WireCodec::Pcm16Le, 48_000),
        "g711_ulaw" => (WireCodec::Ulaw, 8_000),
        "g711_alaw" => (WireCodec::Alaw, 8_000),
        _ => (WireCodec::Pcm16Le, 24_000),
    }
}

pub struct WsAudioPacer {
    ws_send: mpsc::Sender<String>,
    event_seq: Arc<EventSeq>,
    response_id: String,
    item_id: String,
    codec: WireCodec,
    sample_rate: usize,
    frame_samples: usize,
    carry: Vec<f32>,
    payload_buf: Vec<u8>,
    start: Option<Instant>,
    frames_written: u64,
    played_ms: Arc<AtomicU64>,
    last_sample: f32,
    src_position: f64,
}

impl WsAudioPacer {
    pub fn start(
        ws_send: mpsc::Sender<String>,
        event_seq: Arc<EventSeq>,
        played_ms: Arc<AtomicU64>,
        format: &str,
        response_id: &str,
        item_id: &str,
    ) -> Self {
        let (codec, sample_rate) = wire_for(format);
        let frame_samples = sample_rate * FRAME_MS / 1000;
        let payload_bytes = match codec {
            WireCodec::Pcm16Le => frame_samples * 2,
            WireCodec::Ulaw | WireCodec::Alaw => frame_samples,
        };
        Self {
            ws_send,
            event_seq,
            response_id: response_id.to_string(),
            item_id: item_id.to_string(),
            codec,
            sample_rate,
            frame_samples,
            carry: Vec::with_capacity(frame_samples * 4),
            payload_buf: vec![0u8; payload_bytes],
            start: None,
            frames_written: 0,
            played_ms,
            last_sample: 0.0,
            src_position: 0.0,
        }
    }

    pub async fn play(&mut self, audio_24k: MonoF32At24k) -> Result<(), OutboundPushError> {
        if audio_24k.is_empty() {
            return Ok(());
        }
        let samples_24k = audio_24k.into_vec();
        let resampled: Vec<f32> = if self.sample_rate == KOKORO_HZ {
            samples_24k
        } else {
            self.linear_resample(&samples_24k)
        };
        self.carry.extend_from_slice(&resampled);
        while self.carry.len() >= self.frame_samples {
            self.emit_one_frame()
                .await
                .map_err(OutboundPushError::Other)?;
        }
        Ok(())
    }

    pub async fn flush(&mut self) -> Result<()> {
        if self.carry.is_empty() {
            return Ok(());
        }

        while self.carry.len() < self.frame_samples {
            self.carry.push(0.0);
        }
        self.emit_one_frame().await
    }

    async fn emit_one_frame(&mut self) -> Result<()> {
        let frame: Vec<f32> = self.carry.drain(..self.frame_samples).collect();
        match self.codec {
            WireCodec::Pcm16Le => {
                for (i, s) in frame.iter().enumerate() {
                    let v = (s.clamp(-1.0, 1.0) * 32_767.0) as i16;
                    self.payload_buf[i * 2] = v as u8;
                    self.payload_buf[i * 2 + 1] = (v >> 8) as u8;
                }
            }
            WireCodec::Ulaw => {
                for (i, s) in frame.iter().enumerate() {
                    let v = (s.clamp(-1.0, 1.0) * 32_767.0) as i16;
                    self.payload_buf[i] = g711::ulaw_encode_sample(v);
                }
            }
            WireCodec::Alaw => {
                for (i, s) in frame.iter().enumerate() {
                    let v = (s.clamp(-1.0, 1.0) * 32_767.0) as i16;
                    self.payload_buf[i] = g711::alaw_encode_sample(v);
                }
            }
        }
        let event = json!({
            "event_id": self.event_seq.next_id().as_str(),
            "type": "response.output_audio.delta",
            "response_id": self.response_id,
            "item_id": self.item_id,
            "output_index": OUTPUT_INDEX,
            "content_index": CONTENT_INDEX,
            "delta": B64.encode(&self.payload_buf),
        });
        let text = match serde_json::to_string(&event) {
            Ok(t) => t,
            Err(err) => {
                warn!(error = %err, "audio.delta json serialize failed");
                return Ok(());
            }
        };
        if let Err(err) = self.ws_send.send(text).await {
            warn!(error = %err, "ws writer dropped while sending audio.delta");
            return Ok(());
        }
        let start = *self.start.get_or_insert_with(Instant::now);
        self.frames_written += 1;
        self.played_ms
            .store(self.frames_written * FRAME_MS as u64, Ordering::Release);
        let target = start + Duration::from_millis(FRAME_MS as u64) * self.frames_written as u32;
        let now = Instant::now();
        if target > now {
            tokio::time::sleep(target - now).await;
        }
        Ok(())
    }

    fn linear_resample(&mut self, src: &[f32]) -> Vec<f32> {
        if src.is_empty() {
            return Vec::new();
        }
        let ratio = self.sample_rate as f64 / KOKORO_HZ as f64;
        let out_len = ((src.len() as f64) * ratio).ceil() as usize;
        let mut out = Vec::with_capacity(out_len);
        let mut pos = self.src_position;
        let step = 1.0 / ratio;
        let mut last = self.last_sample;
        while pos < src.len() as f64 {
            let lo = pos.floor() as usize;
            let hi = (lo + 1).min(src.len() - 1);
            let frac = (pos - lo as f64) as f32;
            let s = if lo >= src.len() {
                last
            } else {
                let a = src[lo];
                let b = src[hi];
                a + (b - a) * frac
            };
            out.push(s);
            last = s;
            pos += step;
        }
        self.src_position = pos - src.len() as f64;
        self.last_sample = last;
        out
    }
}

pub enum AudioPacer {
    Webrtc(super::audio_out::OutboundPacer),
    WebSocket(WsAudioPacer),
}

impl AudioPacer {
    pub async fn play(&mut self, audio_24k: MonoF32At24k) -> Result<(), OutboundPushError> {
        match self {
            AudioPacer::Webrtc(p) => p.play(audio_24k).await,
            AudioPacer::WebSocket(p) => p.play(audio_24k).await,
        }
    }

    pub async fn flush(&mut self) -> Result<()> {
        match self {
            AudioPacer::Webrtc(p) => p.flush().await,
            AudioPacer::WebSocket(p) => p.flush().await,
        }
    }
}
